Single Responsibility Principle Explained (#287)
reflectoring/reflectoring.github.io: _posts/2020-02-09-single-responsibility-principle.md

    ```
    class BufferedLogger implements Logger {
      private final Logger delegate;
      private final ExecutorService backgroundWorker;
      private final BlockingQueue<Statement> buffer;

      BufferedLogger(Logger delegate) {
        this.delegate = delegate;
        this.backgroundWorker = newSingleThreadExecutor();
        this.buffer = new ArrayBlockingQueue<>(100);
        backgroundWorker.execute(this::writeStatementsInBackground);
      }
      
      @Override public void log(Statement s) {
        putUninterruptibly(buffer, s);
      }

      private void writeStatementsInBackground() {
        while (true) {
          Statement s = takeUninterruptibly(buffer);
          delegate.log(s);
        }
      }
    }
    ```
   `BufferedLogger` is an [active object](https://en.wikipedia.org/wiki/Active_object). Method `writeStatementsInBackground` itself has control flow responsibility.
 
   In a distributed system, examples of services with this type of responsibility could be a proxy, a load balancer, or a service transparently caching responses from or buffering requests to some other service.

#### How small or large a responsibility should be?

> I hope the examples above give some more grounded sense of what a responsibility of a method, class, module, or service could be. However, they still provide no actionable guidance on how finely you should chop responsibilities between the components of your system. For example:

Author: Used

 - Should conversion from XML to JSON be a responsibility of a single method (or a class), or it should be split between two methods: one translates XML into a tree, and another serializes a tree into JSON? Or these should be separate methods belonging to a single class?
 - Should individual types of interactions with an external service (such as different RPC operations) be responsibilities of separate classes,  or they should all belong to a single class? Or, perhaps, interactions should be grouped, such as read operations going to one class and write operations going to another?

>  - How small parts of control flow should be encapsulated in individual methods?

Author: Removed this sentence

>  - What kinds of responsibilities should be split between different (micro)services?

Author: Thanks, applied this suggestion

> Uncle Bob Martin (who first proposed Single Responsibility Principle) suggests that components should be broken down until each one has only one *reason to change*. However,  to me, this criterion still doesn't feel very instructive. Consider the `processTransaction` method above. There may be many reasons to change it:

Reviewer: When referring to a method, add the parentheses. This makes it clear at a glance that you're talking about a method and not a variable name. I.e. `processTransaction()` instead of `processTransaction`.

Author: Fixed

 - Increment counters of normal and fraud transactions to gather statistics.
 - Enrich or reformat the logging statement.
 - Wrap sending an alert into error-handling `try-catch` and log a failure to send an alert.
 
Does this mean that `processTransaction` method is too large and we should split it further into smaller methods? According to Uncle Bob, [we probably should](https://sites.google.com/site/unclebobconsultingllc/one-thing-extract-till-you-drop), but many other people may think that `processTransaction` is already small enough.

> Let's return to the purpose of using Single Responsibility Principle. Obviously, it's to improve the overall quality of the codebase and of its production behavior (Carlo Pescio calls these [artifact space and run-time space](http://www.carlopescio.com/2010/06/notes-on-software-design-chapter-6.html), respectively).

Reviewer: You're introducing codebase / production behavior (artifact space/run-time space) here, but aren't using it further down from what I see. In that case, we should remove these new concepts so as not to overburden the reader.

Author: Now used, in Reusability section

> So, what will utimately help us to apply Single Responsibility Principle effectively is **more clarity into how using the principle affects the quality of the code and the running application**. The optimal scope of the responsibility for a component highly depends on the context:

Reviewer: The sentence "more clarity into how using the principle affects the quality of the code and the running application" takes a lot of my cognitive resources to process :). When you go over your text again correcting things, please try to think of ways how to make sentences like this one easier, perhaps by breaking them up into two sentences.

Author: Rephrased to "So, what will ultimately help us to apply the Single Responsibility Principle effectively is **making clearer for ourselves how SRP affects the quality of the code and the running application**."

It is possible, however, to take the separation of concerns so far that it might become harder to understand the logic. Returning to the `processTransaction()` example, consider the following way of implementing it:

```
class TransactionProcessor {
    private final TransactionInstrumentation instrumentation;
    
    ...
    
    void processTransaction(Transaction t) {
      if (isFraud(t)) {
        instrumentation.detectedFraud(t);
      }
    }
}

class TransactionInstrumentation {
  private final Logger logger;
  private final AlertingService alertingService;

  ...

  void detectedFraud(Transaction t) {
    logger.log("Detected fraud transaction {}", t);
    alertingService.sendAlert(new FraudTransactionAlert(t));
  }
}
```
This approach is not unreasonable. Compared to the original version, it aids the *flexibility* and the *testability* of the code, as we will discuss below in this article. (In fact, I took the idea directly from the excellent article about [domain-oriented observability](https://martinfowler.com/articles/domain-oriented-observability.html) by Pete Hodgson.)

> On the other hand, it smears the logic so thin across multiple classes and methods that it would take longer to learn it than the original, at least for me. [1]

Reviewer: The blog doesn't support foot notes. It does support "notice boxes" though (search for `<div class="notice success">` in the blog codebase). Perhaps that's an alternative.

Author: Notice boxes emphasize. I want to move this part out of the article flow. Maybe asterisk with `<abbr>`-like popup?

  ...

  void detectedFraud(Transaction t) {
    logger.log("Detected fraud transaction {}", t);
    alertingService.sendAlert(new FraudTransactionAlert(t));
  }
}
```
We extracted the observation part of the logic into a separate `TransactionInstrumentation` class. This approach is not unreasonable. Compared to the original version, it aids the *flexibility* and the *testability* of the code, as we will discuss below in this article. (In fact, I took the idea directly from the excellent article about [domain-oriented observability](https://martinfowler.com/articles/domain-oriented-observability.html) by Pete Hodgson.)

On the other hand, it smears the logic so thin across multiple classes and methods that it would take longer to learn it than the original, at least for me. [1]

Extracting responsibilities into separate modules or services (rather than just classes) doesn't help to further improve understandability per se, however it may help with other qualities related the learning curve: the *discoverability* of the functionality (for example, through service API discovery) and the *operability* of the system, which we will discuss below.

> **Understandability itself is somewhat less important when we work on the code alone, rather than in a team.** But don't abuse this - we tend to underestimate how quickly we forget the details of the code on which we worked just a little while ago and how hard it could be to relearn them :)

Author: My sense of typography tells me that there shouldn't be a full stop after a smile (given there is no opening paren corresponding to it). See https://www.quora.com/Does-one-need-to-put-a-full-stop-after-a-smiley-at-the-end-of-a-sentence

### Flexibility
We can easily combine independent components (via separate *control flow* components) in different ways for different purposes, or depending on configuration. Let's take `TransactionProcessor` again:
```
class TransactionProcessor {
  private final AlertingService alertingService;

  ...
    
  void processTransaction(Transaction t) {
    if (isFraud(t)) {
      logger.log("Detected fraud transaction {}", t);
      alertingService.sendAlert(new FraudTransactionAlert(t));
    }
  }

  private boolean isFraud(Transaction t) { ... }
}
```

> To allow the operators of the system to disable alerting, we can add a `NoOpAlertingService` and make it configurable for `TransactionProcessor` via dependency injection. On the other hand, if `sendAlert()` responsibility was not separated out into the `AlertingService` interface, but rather was just a method in `TransactionProcessor`, to make alerting configurable we would have to add a boolean field `sendAlerts` to the class.

Author: Applied

> Imagine now that we want to analyze historical transactions in a framework like Spark. Since `isFraud()` method (that is, the fraud detection responsibility) is a part of `TransactionProcessor`, we must use this class for batch processing. If online and batch processing require different initialization logic, it would go into different constructors of `TransactionProcessor`. On the other hand, if fraud detection was a sole concern of a separate `FraudDetection` class, that would prevent `TransactionProcessor` from swelling.

Author: Applied, but changed the last sentence to "On the other hand, if fraud detection was a concern of a separate `FraudDetection` class, we could prevent `TransactionProcessor` from swelling."

We can notice a pattern: **it is still possible to support different use cases and configuration for a component with multiple responsibilities, but only by increasing the size and the complexity of the component itself**, like adding flags and conditional logic. Little by little, this is how [big ball of mud](https://en.wikipedia.org/wiki/Big_ball_of_mud) systems (aka monoliths), and [runaway methods](https://michaelfeathers.typepad.com/michael_feathers_blog/2012/09/runaway-methods.html) and classes emerge. When each component has a single responsibility, we can keep the complexity of any single one of them limited.

> What about the "lean" approach of splitting responsibilities only when we actually need to make them configurable? I think this is a good strategy, if applied with moderation. It is similar to the Martin Fowler's idea of [preparatory refactoring](https://www.martinfowler.com/articles/preparatory-refactoring-example.html). Keep in mind, however, that **if we don't keep responsibilities separate from early on, they may "fuse" into one another as the code evolves, so it might take much more effort to split them apart further down the road.** To do this, we might also need to spend time relearning the workings of the code in more detail than we would otherwise like to.

Author: Applied

### Reusability

> **When a component has a single, narrow responsibility, it's easier to reuse it.** `FraudDetection` class from the previous section is an example of this: we could reuse it in online processing and batch processing components. To do this in the *artifact space*, we could pull it into a shared library. Another direction is to move fraud detection into a separate microservice, which we can think about as reusability in the *runtime space*. The `FraudDetection` class within our application will then turn from having business logic responsibility to do external integration with the new service.

Reviewer: ```suggestion
**When a component has a single, narrow responsibility, it's easier to reuse it.** The `FraudDetection` class from the previous section is an example of this: we could reuse it in online processing *and* batch processing components. To do this in the *artifact space*, we could pull it into a shared library. Another direction is to move fraud detection into a separate microservice, which we can think about as reusability in the *runtime space*. The `FraudDetection` class within our application will then turn from having business logic responsibility to do external integration with the new service.
```

Author: Applied

> **Most methods with a narrow responsibility shouldn't have side effects and shouldn't depend on the state of the class**, which enables sharing and calling them from any place. In other words, Single Responsibility Principle nudges us toward [functional programming](https://en.wikipedia.org/wiki/Functional_programming) style.

Reviewer: ```suggestion
**Most methods with a narrow responsibility shouldn't have side effects and shouldn't depend on the state of the class**, which enables sharing and calling them from any place. In other words, the Single Responsibility Principle nudges us toward a [functional programming](https://en.wikipedia.org/wiki/Functional_programming) style.
```

Author: Applied

***
Pro tip: thinking about responsibilities helps to notice [unrelated subproblems](https://learning.oreilly.com/library/view/the-art-of/9781449318482/ch10.html) hiding in our methods and classes. When we extract them, we can then see opportunities to reuse them in other places. Moving unrelated subproblems out of the way keeps a component at the [single level or abstraction](http://principles-wiki.net/principles:single_level_of_abstraction), which makes easier to understand the logic of the component.

> ***

Reviewer: This is a candidate for an info box, I'd say. Would be good to have a couple to break up the text.

Author: Agreed

### Testability

> **It's easier to test methods and classes with focused, independent concerns.** This is what [Humble Object](http://xunitpatterns.com/Humble%20Object.html) pattern is all about. Let's continue playing with `TransactionProcessor`:

Reviewer: ```suggestion
**It's easier to test methods and classes with focused, independent concerns.** This is what the [Humble Object](http://xunitpatterns.com/Humble%20Object.html) pattern is all about. Let's continue playing with `TransactionProcessor`:
```

Author: Applied

```
class TransactionProcessor {
    
  void processTransaction(Transaction t) {
    boolean isFraud;
    // logic detecting that the transaction is fraud, many

>     // line of code omitted

Reviewer: ```suggestion
    // lines of code omitted
```

Author: Applied

    ...
    
    if (isFraud) {
      logger.log("Detected fraud transaction {}", t);
      alertingService.sendAlert(new FraudTransactionAlert(t));
    }
  }
}
```

> In this variant, there is no separate `isFraud()` method. `processTransaction()` conflates the fraud detection and the reporting logic. Then, to test the fraud detection, we either need to mock the `alertingService`, which pollutes the test code with boilerplate, or to intercept and check the logging output, which is also cumbersome, and may also hinder our ability to execute tests in parallel.

Reviewer: ```suggestion
In this variant, there is no separate `isFraud()` method. `processTransaction()` combines the fraud detection and the reporting logic. To test the fraud detection logic, we don't need the alerting logic. So, we either need to mock the `alertingService`, which pollutes the test code with boilerplate, or we intercept and check the logging output, which is also cumbersome, and may even hinder our ability to execute tests in parallel.
```

Author: Why conflates -> combines? I want to have a negative connotation.

Author: Also rephrased this part separately, please review again.

> It's simpler to test a separate `isFraud()` method. But we would still need to construct a `TransactionProcessor` object and to pass some dummy `Logger` and `AlertingService` objects into it. So it's still easier to test the variant with `FraudDetection` class. Notice that to test the intermediate version without separate `FraudDetection` class, we often find ourselves changing the visibility of the method under test (`isFraud()`, in this example) from private to default (package-private). **Use changing visibility of a method and `@VisibleForTesting` annotation as clues to think if we should split the responsibilities of the enclosing class.**

Reviewer: ```suggestion
It's simpler to test a separate `isFraud()` method. But we would still need to construct a `TransactionProcessor` object and to pass some dummy `Logger` and `AlertingService` objects into it. So it's even easier to test the variant with `FraudDetection` class. Notice that to test the intermediate version without separate `FraudDetection` class, we often find ourselves changing the visibility of the method under test (`isFraud()`, in this example) from private to default (package-private). **We should use changing the visibility of a method and the `@VisibleForTesting` annotation as clues to think about splitting up responsibilities into multiple classes.**
```

Author: Applied still -> even, not sure about "We should use changing", to me it sounds a little odd. There are other places in text with command language "Note" or "Beware", but we don't replace them all with "We should note" and "we should be aware".

> The article about domain-oriented observability referenced above also explains how extracting observability into a separate class (like `TransactionInstrumentation`) [enables clearer, more focused tests](https://martinfowler.com/articles/domain-oriented-observability.html#DomainProbesEnableCleanerMore-focusedTests).

Reviewer: Which article? Can you link it again?

Author: Shortened this sentence without a reference to the article, just link (the article is under the link).

This article explains the Single Responsibility Principle (SRP): what does it practically mean, when and how to apply it.

>  - [What does the Single Responsibility Principle say?](#what-does-the-single-responsibility-principle-say)

Reviewer: We don't need a table of contents, it's generated automatically in the side bar.

Author: Removed

Alternatively, to test the fraud detection logic in the example above, we could intercept and check the logging output. However, this is also cumbersome, and it hinders the ability to execute tests in parallel.

It's simpler to test a separate `isFraud()` method. But we would still need to construct a `TransactionProcessor` object and to pass some dummy `Logger` and `AlertingService` objects into it. So it's even easier to test the variant with the `FraudDetection` class. Notice that to test the intermediate version without separate `FraudDetection` class, we often find ourselves changing the visibility of the method under test (`isFraud()`, in this example) from private to default (package-private). **Use changing visibility of a method and `@VisibleForTesting` annotation as clues to think about whether it's better to split the responsibilities of the enclosing class.**

Pete Hodgson also explains how extracting observability into a separate class (like `TransactionInstrumentation`) [enables clearer, more focused tests](https://martinfowler.com/articles/domain-oriented-observability.html#DomainProbesEnableCleanerMore-focusedTests).

> In contrast to methods and classes, **smaller services complicate the local setup for integration testing.** [Docker Compose](https://docs.docker.com/compose/) is a godsend, but it doesn't solve the problem fully.

Author: I use term "service" in the sense of separate executable throughout this article, nevertheless, replaced with "(micro)services" in this particular case.

**When microservices have focused responsibilities, it also helps to investigate incidents.** If we monitor the request success rates and health status of each service, and see that one service which connects to a particular database is failing or unavailable, we may assume that the root problem of lies in this database rather than any other part of the system.

However, despite of the advantages of finer-grained monitoring and scaling, **splitting responsibilities between smaller services generally increases the burden of operating the system.** Smaller services mean more work:
- Setting up and operating intermediate message queues (like Kafka) between the services.
 - DevOps: setting up and managing separate delivery pipelines, monitoring, configuration, machine and container images.
 - Deployment and orchestration: Kubernetes doesn't fully alleviate it.
 - To ensure [rollback safety](https://aws.amazon.com/builders-library/ensuring-rollback-safety-during-deployments/), the deployments should be multi-phase, shared state and messages sent between services should be versioned.

### Reliability

> Reliability is the first software quality in the list that we mostly hurt, not aid when we split smaller responsibilities between the components. Although, if engineered properly, microservices can imporve availability: when one service is sick, others might still serve something for the users, the inherent [fallability of distributed systems](https://en.wikipedia.org/wiki/Fallacies_of_distributed_computing) hits harder. Also, "if engineered properly" is an important caveat :)

Reviewer: Why do we hurt reliability?

Author: Expanded and rephrased this part a bit, mentioned network partition and degradation.

Discussing the pros and cons of microservices is not the main goal of this article, but there are plenty of good materials on this topic, the reliability aspect in particular: [1](https://www.martinfowler.com/articles/microservice-trade-offs.html), [2](https://dwmkerr.com/the-death-of-microservice-madness-in-2018/), [3](https://developer.ibm.com/technologies/microservices/articles/challenges-and-benefits-of-the-microservice-architectural-style-part-1/).

### Code Size
Smaller responsibility of each component means that there are more components in total in the system.

Each method needs a signature declaration. Each class needs constructors, static factory methods, field declarations, and other ceremony. Each module needs a separate configuration class and a depedendency injection setup. Each service needs separate configuration files, startup scripts, CI/CD/orchestration infrastructure, and so on.

> Therefore, **the more focused responsibilities of the components we make, the more code we will need to write.** This impacts the long-term maintainability much less than all the qualities discussed above: undertandability, flexibility, reusability, etc. However, it means that **it takes more time and effort to develop the first version of the application with finely separated responsibilites than with larger components**, especially if thorough testing is not the first priority as well.

Reviewer: Yes! Building clean software takes longer initially but will save time later! I said similar things in my book :)

This may sound overwhelming, but of course this analysis shouldn't be done for each and every method and class in separation, but instead done infrequently to establish a guideline on the project, or just to train our intuition.

On the level of the distributed system, the tradeoff is much less in favor of extracting (micro)services with more narrow responsibilities: discoverability, flexibility, reusability, and observability improves, but testability, operability, reliability, and performance mostly decline. On the other hand, Single Reponsibility Principle probably shouldn't be the first thing to consider when sizing microservices. Most people in the industry think that it's more important to follow the [team boundaries](https://martinfowler.com/articles/microservices.html#OrganizedAroundBusinessCapabilities), [bounded contexts](https://martinfowler.com/bliki/BoundedContext.html), and [aggregates](https://www.martinfowler.com/bliki/DDD_Aggregate.html) (the last two are concepts from Domain-Driven Design).

<hr>

> P. S. I expolore the idea of analyzing software design practices and principles through the lenses of distinct software qualities such understandability, testability, performance, and so on in the [Software Design](https://en.wikiversity.org/wiki/Software_Design) project on Wikiversity.

Reviewer: You might want to put this in your author bio. Add yourself to `authors.yml`, with a 180x180 profile picture and put this text into the `bio` field. It supports HTML links.

I added a header to this file, so that we can preview the article now. In the header, reference your author name from `authors.yml`.

Author: Added

> [1] For code understandability, the crossing point might actually depend on  individual cognitive features of the developer: attention, the tendency to control things more tightly or more loosely, the habits of working with code. I suspect this may the crux of the debate around the Robert Martin's hard "[extract till you drop](https://sites.google.com/site/unclebobconsultingllc/one-thing-extract-till-you-drop)" recommendation.

Reviewer: The blog engine doesn't support foot notes. Can you put this into a notice box in the section this is references? This would also help to loosen up the text a bit :)

Author: Removed this part

This article explains the Single Responsibility Principle (SRP): what does it practically mean, when and how to apply it.

> ### What does Single Responsibility Principle say?

Author: Converted

Single Responsibility Principle may feel a bit vague at first. Let's try to deconstruct it and look at what does it actually mean.

Single Responsibility Principle applies to the software that we develop on different levels: to methods, classes, module, and services (collectively, I'll call all these things *components* later in this article). So, on each level, we can phrase the principle like the following:

>  - Each method (function) should have a single responsibility.

Author: Rephrased this part

 - Each class (object) should have a single responsibility.
 - Each module (subsystem) should have a single responsibility.
 - Each service should have a single responsibility.

These phrases are now more concrete, but they still don't explain what a *responsibility* is and *how small or large a responsibility should be* for each particular method, class, module, or service.

> #### Types of responsibilities

Author: Converted

Instead of defining a *responsibility* in abstract terms, it may be more intuitive to list the actual types of responsibilities. Here are some examples:

> 1. **Business logic**, for example, extracting a phone number from text, converting XML document into JSON, or classifying a money transaction as fraud. On the level of methods, such a responsibility presents itself as a single time the logic is applied, for example, extract a phone number from one given text. On the level of classes and above, it's *knowing how to do* (or encapsulating) the business function: for example, a class knowing how to convert XML documents into JSON, or a service encapsulating the detection of fraud transactions.

Author: > I'd make each of these bullet points a separate subsection to give the text a little more white space.

Separated with whitespace lines

> I don't understand this sentence: "On the level of methods, such a responsibility presents itself as a single time the logic is applied, for example, extract a phone number from one given text." What does "applying a single time" mean?
>
> Please rephrase in easier words.

Removed that sentence and a similar sentence in the next section (external integration).

3. **Data**: a profile of a person on a website, a JSON document, a message. Obviously, embodying a piece of data could only be a responsibility of a class (object), but not of a method, module, or service. A specific kind of data is **configuration**: a collection of parameters for some other method, class, or system.

> 4. A piece of application's **control flow, execution, or data flow** model.  One example are methods that glue together calls to other methods:

Reviewer: ```suggestion
4. A piece of application's **control flow, execution, or data flow** model.  An example of this responsibility is a method that orchestrates calls to components that each have other responsibilities:
```

Author: Applied

    ```
    void processTransaction(Transaction t) {
      if (isFraud(t)) { // Business logic
        // External integration: logging
        logger.info("Detected fraud transaction {}", t);
        // Integration with external service
        alertingService.sendAlert(new FraudTransactionAlert(t));
      }
    }
    ```
    Another example may be a `BufferedLogger` class which buffers logging statements in memory and manages a separate background thread that takes statements from the buffer and writes them to actual external logger:
    ```

>     class BufferedLogger implements Logger {

Reviewer: I find the `BufferedLogger` example too complex here. 30 lines of code of which only the method `writeStatementsInBackground` is really relevant. And to understand this method, you need to know what `putInterrubtibly` means and does. Can you find a simpler example (with fewer lines of code) of an Active Object that does not introduce new concepts?

      private final Logger delegate;
      private final ExecutorService backgroundWorker;
      private final BlockingQueue<Statement> buffer;

      BufferedLogger(Logger delegate) {
        this.delegate = delegate;
        this.backgroundWorker = newSingleThreadExecutor();
        this.buffer = new ArrayBlockingQueue<>(100);
        backgroundWorker.execute(this::writeStatementsInBackground);
      }
      
      @Override public void log(Statement s) {
        putUninterruptibly(buffer, s);
      }

      private void writeStatementsInBackground() {
        while (true) {
          Statement s = takeUninterruptibly(buffer);
          delegate.log(s);
        }
      }
    }
    ```

>    `BufferedLogger` is an [active object](https://en.wikipedia.org/wiki/Active_object). Method `writeStatementsInBackground` itself has control flow responsibility.

Reviewer: You introduce the term "Active Object" without explaining it. Does it help in this context to know that `BufferedLogger` is an active object? Of not, leave this sentence away, if yes, please add a sentence explaining what it is.

Author: Removed sentence.

>    In a distributed system, examples of services with this type of responsibility could be a proxy, a load balancer, or a service transparently caching responses from or buffering requests to some other service.

Reviewer: With "with this type of responsibility" you mean Active Objects? Or do you mean objects with control flow responsibility? Please make this clearer.

Author: Replaced with "examples of services with a control or data flow responsibility could be ..."

> (These types are derived from Adam Warski's classification of objects in applications which he coined in his thought-provoking post about [dependency injection in Scala](https://blog.softwaremill.com/zio-environment-meets-constructor-based-dependency-injection-6a13de6e000).)

Reviewer: I'd move this sentence and the reference to the source up into the intro of "Types of Responsibilities" section. I was confused at first if this sentence belonged to the last example or to the whole section.

Author: Moved up.

#### How small or large a responsibility should be?
I hope the examples above give some more grounded sense of what a responsibility of a method, class, module, or service could be. However, they still provide no actionable guidance on how finely you should chop responsibilities between the components of your system. For example:

 - Should conversion from XML to JSON be a responsibility of a single method (or a class), or should it be split between two methods: one translates XML into a tree, and another serializes a tree into JSON? Or should these be separate methods belonging to a single class?
 - Should individual types of interactions with an external service (such as different RPC operations) be responsibilities of separate classes,  or they should all belong to a single class? Or, perhaps, should interactions be grouped, such as read operations going to one class and write operations going to another?

>  - How small should the parts of a control flow should be encapsulated in individual methods?

Author: Doesn't seem this correction is valid to me.

Author: Irrelevant now, removed that sentence

 - What kinds of responsibilities should be split between different (micro)services?

Uncle Bob Martin (who first proposed Single Responsibility Principle) suggests that components should be broken down until each one has only one *reason to change*. However,  to me, this criterion still doesn't feel very instructive. Consider the `processTransaction()` method above. There may be many reasons to change it:
 - Add a counter to gather statistics of normal and fraud transactions.
 - Enrich or reformat the logging statement.
 - Add error handling that logs when `sendAlert()` throws an exception.
 

> Does this mean that `processTransaction()` is too large and we should split it further into smaller methods? According to Uncle Bob, [we probably should](https://sites.google.com/site/unclebobconsultingllc/one-thing-extract-till-you-drop), but many other people may think that `processTransaction` is already small enough.

Author: Replaced with "that the `processTransaction()` method ..."
